---
title: "in silico RNA mixing"
author: "Xueyi Dong"
date: "16/12/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Arrange sample information

We aim to see whether *in silico* mixture samples look similar enough to the ones obtained via physical mixing of RNA in the lab. The mixture proportions are 25:75, 50:50, 75:25.

For each sample, we used seqtk to subsample X, X and X reads from pure samples across all lanes, and combine the reads to create the in silico mixture, X reads each. Then, we use the similar mapping and counting strategy comparing to the paper.

## Alignment and counting

We aligned the reads to reference transcriptome using minimap2 and quantified them using salmon. Details are described in the main text of the paper. 

## Data pre-processing

read in count data from txt files using **edgeR** `catchSalmon`.
```{r lib}
library(edgeR)
library(limma)
```


```{r echo=FALSE}
# DIR="/vast/scratch/users/dong.x/long_read_benchmark/pilot_ONT/salmon"
DIR = "/stornext/General/data/user_managed/grpu_mritchie_1/XueyiDong/long_read_benchmark/pilot_ONT/salmon"
counts <- catchSalmon(file.path(DIR, list.files(DIR))[c(1:10, 13:18)])
```






```{r DGE}
x <- DGEList(counts = counts$counts, genes = counts$annotation)
colnames(x) <- strsplit2(colnames(x), "/")[,11]
dim(x)

x$samples$rep <- rep(c("rep1", "rep2", "rep1", "rep2"), c(5, 5, 3, 3))
x$samples$prop <- c(rep(c("000", "025", "050", "075", "100"), 2), 
                    rep(c("025", "050", "075"), 2))
x$samples$source <- c(rep("seq", 10), rep("mix", 6))
x$samples$group <- paste(x$samples$source, x$samples$prop, sep="_")
x$samples
```

Filter out lowly expressed genes

```{r filter}
keep.exprs <- filterByExpr(x, group=x$samples$group)
x <- x[keep.exprs,, keep.lib.sizes=FALSE]
dim(x)
```

Normalization using TMM

```{r norm}
x <- calcNormFactors(x, method = "TMM")
x$samples$norm.factors
```

## MDS plot

```{r}
lcpm <- cpm(x, log=TRUE)
plotMDS(lcpm)
library(RColorBrewer)
col <- brewer.pal(5, "Set1")[c(1, 5, 4, 3, 2)]
pdf("MDS_pilot.pdf", height = 5, width = 8)
plotMDS(lcpm, 
        pch = rep(c(1, 2), c(10, 6)),
        col = col[c(rep(1:5, 2), rep(2:4, 2))], 
        cex.main = 1.5, cex.lab = 1.5, cex.axis = 1.5, cex = 1.5)
legend("bottomleft", c("000", "025", "050", "075", "100", "lab-derived samples", "in silico mixture"),
        text.col = c(col, "black", "black"), cex = 1.25, bty = "n",
       pch = c(rep(NA, 5), 1, 2))
dev.off()
pdf("MDS_pilot.pdf", height = 5, width = 8)
plotMDS(lcpm,
        col = col[c(rep(1:5, 2), rep(2:4, 2))], 
        cex.main = 1.5, cex.lab = 1.5, cex.axis = 1.5, cex = 1.5)
dev.off()
pdf("MDS_pilot_rep1.pdf", height = 5, width = 8)
plotMDS(lcpm[, c(1:5, 11:13)],
        pch = rep(c(1, 2), c(5, 3)),
        col = col[c(1:5, 2:4)],
        ylim = c(-0.8, 1.4),
        cex.main = 1.5, cex.lab = 1.5, cex = 1.5)
legend("bottomleft", c("000", "025", "050", "075", "100", "lab-derived samples", "in silico mixture"), 
        text.col = c(col, "black", "black"), cex = 1.25, bty = "n",
       pch = c(rep(NA, 5), 1, 2))
dev.off()

```

## Count correlation

heatmap
```{r}
cormat <- cor(x$counts[, c(1:5, 11:13)], method = "spearman")
library(pheatmap)
anno <- x$samples[c(1:5, 11:13), c("prop", "source")]
colnames(anno) <- c("mix_proportion", "source")
anno$source <- ifelse(anno$source == "seq", "lab-derived", "in-silico mixture")
anno_colours = list(
  mix_proportion = c(`000` = col[1], `025` = col[2], `050` = col[3], `075` = col[4], `100` = col[5]),
  source = c(`lab-derived` = "salmon", `in-silico mixture` = "seagreen2")
)
pheatmap(cormat,
         cluster_cols = FALSE,
         cluster_rows = FALSE,
         show_colnames = FALSE,
         show_rownames = FALSE,
         annotation_col = anno,
         annotation_row = anno,
         annotation_colors = anno_colours,
         scale = "none",
         display_numbers = TRUE)

```

```{r}
p25 <- ggplot(as.data.frame(x$counts), aes(x=barcode02_pass, y=`rep1-025`)) +
  stat_bin_hex(bins = 100) +
  annotate(geom = "text", x = 50, y = 30000,
           label = paste0("Spearman's rho =", round(cor(x$counts[,2], x$counts[, 11], method = "spearman"), 3))) +
  labs(x = "lab-derived sample", y = "in silico mixture", title = "Sample 025") +
  scale_fill_viridis() +
  scale_x_continuous(trans = "log10") +
  scale_y_continuous(trans = "log10") +
  theme_bw() +
  theme(text = element_text(size = 16))

p50 <- ggplot(as.data.frame(x$counts), aes(x=barcode03_pass, y=`rep1-050`)) +
  stat_bin_hex(bins = 100) +
  annotate(geom = "text", x = 50, y = 30000,
           label = paste0("Spearman's rho =", round(cor(x$counts[,3], x$counts[, 12], method = "spearman"), 3))) +
  labs(x = "lab-derived sample", y = "in silico mixture", title = "Sample 050") +
  scale_fill_viridis() +
  scale_x_continuous(trans = "log10") +
  scale_y_continuous(trans = "log10") +
  theme_bw() +
  theme(text = element_text(size = 16))

p75 <- ggplot(as.data.frame(x$counts), aes(x=barcode03_pass, y=`rep1-075`)) +
  stat_bin_hex(bins = 100) +
  annotate(geom = "text", x = 50, y = 30000,
           label = paste0("Spearman's rho =", round(cor(x$counts[,4], x$counts[, 13], method = "spearman"), 3))) +
  labs(x = "lab-derived sample", y = "in silico mixture", title = "Sample 075") +
  scale_fill_viridis() +
  scale_x_continuous(trans = "log10") +
  scale_y_continuous(trans = "log10") +
  theme_bw() +
  theme(text = element_text(size = 16))

pdf("countCor.pdf", height = 6, width = 16)
cowplot::plot_grid(p25, p50, p75, ncol = 3)
dev.off()
```

# mixture model analysis
```{r}
Expr000 <- lcpm[,1]
Expr100 <- lcpm[,5]

### in silico mixture
Expr <- lcpm[, 11:13]
out <- fitmixture(Expr, mixprop=c(0.25, 0.5, 0.75))

# Plot true vs estimated log-ratios
pdf("mixModelInSilico.pdf", height = 5, width = 8)
plot(Expr100-Expr000, out$M, xlim=c(-8,8), ylim=c(-8,8), pch=16, cex=0.4,
    xlab="logFC (100 vs 0)", ylab="Estimated logFC (100 vs 0)", main="ONT - in silico mixture")
abline(a=0, b=1, col="red", lty=2, lwd=2)
rho <- cor(Expr100-Expr000, out$M, method = "spearman")
legend("topleft", legend=paste("Spearman cor = ", round(rho, digits=2)), bty="n")
dev.off()

### Lab-derived
Expr <- lcpm[, 2:4]
out <- fitmixture(Expr, mixprop=c(0.25, 0.5, 0.75))

# Plot true vs estimated log-ratios
pdf("mixModelLabMix.pdf", height = 5, width = 8)
plot(Expr100-Expr000, out$M, xlim=c(-8,8), ylim=c(-8,8), pch=16, cex=0.4,
    xlab="logFC (100 vs 0)", ylab="Estimated logFC (100 vs 0)", main="ONT - Lab-derived samples")
abline(a=0, b=1, col="red", lty=2, lwd=2)
rho <- cor(Expr100-Expr000, out$M, method = "spearman")
legend("topleft", legend=paste("Spearman cor = ", round(rho, digits=2)), bty="n")
dev.off()
```

