---
title: "in silico RNA mixing"
author: "Xueyi Dong"
date: "12/02/2019, updated 18/05/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Arrange sample information

We aim to see whether *in silico* mixture samples look similar enough to the ones obtained via physical mixing of RNA in the lab. The mixture proportions are 25:75, 50:50, 75:25.

In the original paper, FASTQ files from the same libraries were merged and aligned to the hg19 build of the human reference genome using the Subread and Subjunc software (version 1.4.6) with default settings.

For each sample, we used seqtk to subsample 10M, 20M and 30M reads from pure samples across all lanes, and combine the reads to create the in silico mixture, 40M reads each. Then, we use the similar mapping and counting strategy comparing to the paper.

Here, we arrange mixed sample information.
```{r mixinfo}
dir <- "/stornext/General/data/user_managed/grpu_mritchie_1/Mixture/MainStudy/in_silico_mix"
mixinfo <- data.frame(
  rep = c(rep("R1", 3), rep("R2", 3), rep("R3", 3)),
  prop = rep(c("025", "050", "075"), 3)
)
mixinfo$file <- paste(mixinfo$rep, "-", mixinfo$prop, "sub.fastq", sep="")
mixinfo$bam <- paste(mixinfo$rep, "-", mixinfo$prop, "sub.bam", sep="")
mixinfo$bampath <- paste(dir, "bam", mixinfo$bam, sep="/")
mixinfo
```

Arrange raw sample counts info

```{r sampleinfo}
sampleinfo <- data.frame(
  rep = c(rep("R1", 5), rep("R2", 5), rep("R3", 5)),
  prop = rep(c("000", "025", "050", "075", "100"), 3)
)
sampleinfo$bam <- paste(sampleinfo$rep, "-", sampleinfo$prop, ".bam", sep="")
sampleinfo$bampath <- paste(dir, "bam", sampleinfo$bam, sep="/")
sampleinfo
```

## Alignment and counting

We merged data from different lanes and performed sampling using shell scripts. Then we perform alignment and feature count to hg19. The hg19 index have already built in the past, so we can directly use it.

```{r subread, eval=FALSE}
library(Rsubread)
# index <- "/wehisan/home/allstaff/d/dong.x/personal/index/hg19/hg19_index"
fc <- featureCounts(c(mixinfo$bampath, sampleinfo$bampath), annot.inbuilt = "hg19", nthreads = 24)
save(fc, file="featureCountsResult.RData")
```

```{r echo=FALSE}
load("featureCountsResult.RData")
```


```{r savefc}
counts <- fc$counts
colnames(counts) <- c(mixinfo$bam, sampleinfo$bam)
```

## Data pre-processing

read in count data from txt files using **edgeR**

```{r lib}
library(edgeR)
library(limma)
```

```{r DGE}
x <- DGEList(counts = counts)
dim(x)
x$samples$rep <- c(mixinfo$rep, sampleinfo$rep)
x$samples$prop <- c(as.character(mixinfo$prop), as.character(sampleinfo$prop))
x$samples$source <- c(rep("mix", 9), rep("seq", 15))
x$samples$group <- paste(x$samples$source, x$samples$prop, sep="_")
x$samples
```

Filter out lowly expressed genes

```{r filter}
keep.exprs <- filterByExpr(x, group=x$samples$group)
x <- x[keep.exprs,, keep.lib.sizes=FALSE]
dim(x)
```

Normalization using TMM

```{r norm}
x <- calcNormFactors(x, method = "TMM")
x$samples$norm.factors
```

## MDS plot

```{r}
lcpm <- cpm(x, log=TRUE)
library(RColorBrewer)
col <- brewer.pal(5, "Set1")[c(1, 5, 4, 3, 2)]
pdf("MDS_pilot.pdf", height = 5, width = 8)
plotMDS(lcpm, 
        pch = rep(c(2, 1), c(9, 15)),
        col = col[c(rep(2:4, 3), rep(1:5, 3))], 
        cex.main = 1.5, cex.lab = 1.5, cex.axis = 1.5, cex = 1.5)
legend("bottomleft", c("000", "025", "050", "075", "100", "lab-derived samples", "in silico mixture"), 
        text.col = c(col, "black", "black"), cex = 1.25, bty = "n",
       pch = c(rep(NA, 5), 1, 2))
dev.off()
```

# Mixture model analysis

```{r}
Expr000 <- rowMeans(lcpm[, x$samples$prop == "000"])
Expr100 <- rowMeans(lcpm[, x$samples$prop == "100"])

# in silico mixture
Expr <- lcpm[, 1:9]
out <- fitmixture(Expr, mixprop=rep(c(0.25, 0.5, 0.75),3))
# Plot true vs estimated log-ratios
pdf("mixModelInSilico.pdf", height = 5, width = 8)
plot(Expr100-Expr000, out$M, xlim=c(-8,8), ylim=c(-8,8), pch=16, cex=0.4,
    xlab="logFC (100 vs 0)", ylab="Estimated logFC (100 vs 0)", main="Illumina - in silico mixture")
abline(a=0, b=1, col="red", lty=2, lwd=2)
rho <- cor(Expr100-Expr000, out$M, method = "spearman")
legend("topleft", legend=paste("Spearman cor = ", round(rho, digits=2)), bty="n")
dev.off()

# 1 sample
TrueY1 <- lcpm[, 10]
TrueY2 <- lcpm[, 14]
Expr <- lcpm[, 1:3]
out <- fitmixture(Expr, mixprop=rep(c(0.25, 0.5, 0.75),1))
# Plot true vs estimated log-ratios
pdf("mixModelInSilico1Sample.pdf", height = 5, width = 8)
plot(TrueY2 - TrueY1, out$M, xlim=c(-8,8), ylim=c(-8,8), pch=16, cex=0.4,
    xlab="logFC (100 vs 0)", ylab="Estimated logFC (100 vs 0)", main="Illumina - in silico mixture (1 sample)")
abline(a=0, b=1, col="red", lty=2, lwd=2)
rho <- cor(TrueY2-TrueY1, out$M, method = "spearman")
legend("topleft", legend=paste("Spearman cor = ", round(rho, digits=2)), bty="n")
dev.off()

### Lab-derived
Expr <- lcpm[, c(11:13,16:18,21:23)]
out <- fitmixture(Expr, mixprop=rep(c(0.25, 0.5, 0.75),3))

# Plot true vs estimated log-ratios
pdf("mixModelLabMix.pdf", height = 5, width = 8)
plot(Expr100- Expr000, out$M, xlim=c(-8,8), ylim=c(-8,8), pch=16, cex=0.4,
    xlab="logFC (100 vs 0)", ylab="Estimated logFC (100 vs 0)", main="Illumina - Lab-derived samples")
abline(a=0, b=1, col="red", lty=2, lwd=2)
rho <- cor(Expr100-Expr000, out$M, method = "spearman")
legend("topleft", legend=paste("Spearman cor = ", round(rho, digits=2)), bty="n")
dev.off()

# 1 sample
TrueY1 <- lcpm[, 10]
TrueY2 <- lcpm[, 14]
Expr <- lcpm[, 11:13]
out <- fitmixture(Expr, mixprop=rep(c(0.25, 0.5, 0.75),1))
# Plot true vs estimated log-ratios
pdf("mixModelLabMix1sample.pdf", height = 5, width = 8)
plot(TrueY2 - TrueY1, out$M, xlim=c(-8,8), ylim=c(-8,8), pch=16, cex=0.4,
    xlab="logFC (100 vs 0)", ylab="Estimated logFC (100 vs 0)", main="Illumina - Lab-derived samples (1 sample)")
abline(a=0, b=1, col="red", lty=2, lwd=2)
rho <- cor(TrueY2-TrueY1, out$M, method = "spearman")
legend("topleft", legend=paste("Spearman cor = ", round(rho, digits=2)), bty="n")
dev.off()
```

